### PR TITLE
[runtime env] URI reference refactor

### PR DESCRIPTION
**Pull request by SongGuyang**
## Why are these changes needed?
- Move the URI reference logic from raylet to agent.
- Redefine the runtime env agent RPC to `CreateRuntimeEnvOrGet` and `DeleteRuntimeEnvIfPossible`
- More details https://github.com/ray-project/ray/issues/21695#issuecomment-1032161528

Future works
- We don't remove the `RuntimeEnvUris` from `RuntimeEnv` protobuf in current PR because gcs also uses those URIs to do GC by runtime_env_manager. We should also clear this.
- Ray client server shouldn't interact with agent directly. Or Ray client server should also decrease the reference count.
- Currently, `WorkerPool::HandleJobStarted` will be called multiple times for one job. So we should make sure this function is idempotent. Can we change this logic and make this function be called only once?

## Related issue number

https://github.com/ray-project/ray/issues/21695

## Checks

- [X] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by SongGuyang** [comment]
@edoakes @architkulkarni @rkooo567 This PR is not completed yet. But I have one question needed to discuss first. I saw that we also create runtime env in client server, but it doesn't delete the URI by RPC. Do we will keep this logic in future? Maybe there will be some corner cases which bring URI leakage(for example, client-server dies before the client job started) . I'm not sure the real effect.
Another thought, can we setup the environment of client-server by the way like job submission? If we can start client-server in an actor, we can avoid proxier communicating with agent.

**Comment by rkooo567** [comment]
I believe that code path is a hack and we should remove it!

**Comment by SongGuyang** [comment]
@architkulkarni @edoakes @rkooo567 This PR is ready to review! I have updated some issues in the `Future works` part of the description. I'd like to discuss about this.

File: src/ray/raylet/worker_pool.cc
Comment (architkulkarni) [inline src/ray/raylet/worker_pool.cc:582]: We currently have `IsRuntimeEnvEmpty()` in `runtime_env_common.h`, does it make sense to use that one?

File: src/ray/raylet/worker_pool.cc
Comment (architkulkarni) [inline src/ray/raylet/worker_pool.cc:737]: I don't have much context here, do you know why HandleJobStarted wasn't called here previously?  Was this a bug?

File: src/ray/raylet/worker_pool.h
Comment (architkulkarni) [inline src/ray/raylet/worker_pool.h:463]:
```suggestion
  /// Some basic information about the worker process.
```

File: src/ray/raylet/worker_pool.h
Comment (architkulkarni) [inline src/ray/raylet/worker_pool.h:841]: Isn't there still value in keeping this? (Maybe moving it above the Increase/DecreaseURIReference methods)?  It's true that the actual adding and deleting of the references is now happening in the agent, but the AddURIReference and DeleteURIReference calls are still coming from this file.  So the information here is still accurate and useful, right?

**Review by architkulkarni — COMMENTED** [review]
Looks great!  Just some minor comments and questions.

- Regarding the `HandleJobStarted` future work item in the PR description, I don't know if it's possible to make it only be called once per job -- @wuisawesome maybe you have some thoughts about this?

- I'm not too familiar with the logic of `num_registered_workers` and `num_starting_workers` so it's hard for me to review that part carefully.  @SongGuyang maybe you could summarize the change, or someone else could review that part?

- It seems like the worker pool is now calling AddURIReference, which creates the runtime env as a side effect.  The name might be a little confusing. Shouldn't these methods be called CreateRuntimeEnvIfNeeded and DeleteRuntimeEnvIfNeeded, and internally the agent adjusts the refcounts if needed?  I don't feel strongly about this though, I'm fine with leaving it.

**Comment by SongGuyang** [comment]
@rkooo567 please help to review the part of `worker_pool`?

File: src/ray/raylet/worker_pool.cc
Comment (wuisawesome) [reply src/ray/raylet/worker_pool.cc:737]: +1 to Archit's comment. Either way, can we get a unit test to ensure the ref counting behavior is correct and stays correct?

**Comment by SongGuyang** [comment]
> * It seems like the worker pool is now calling AddURIReference, which creates the runtime env as a side effect.  The name might be a little confusing. Shouldn't these methods be called CreateRuntimeEnvIfNeeded and DeleteRuntimeEnvIfNeeded, and internally the agent adjusts the refcounts if needed?  I don't feel strongly about this though, I'm fine with leaving it.

@edoakes @rkooo567 what do you think about the RPC name? `CreateRuntimeEnvIfNeeded` and `DeleteRuntimeEnvIfNeeded` seem also good for me, but there is no meaning of `reference count`. I'm a little entangled with the names. 😅

**Comment by rkooo567** [comment]
I think it is okay not to expose ref count to worker pool, so “ifNeeded” sounds good to me?

**Comment by rkooo567** [comment]
And I will review it today!

File: src/ray/raylet/worker_pool.cc
Comment (SongGuyang) [reply src/ray/raylet/worker_pool.cc:737]: `HandleJobStarted` only do two things: First, add job to `all_jobs_ `. Second, install runtime env eagerly if needed.

Before current PR:
`HandleJobStarted` will be triggered by the publishing channel of job. So here, in `RegisterDriver`, it's ok to only add job to `all_jobs_`.

In current PR:
To guarantee the create runtime env RPC only be called once, I have modified the logic of  `HandleJobStarted`. `HandleJobStarted` will check if the job already been added. So, I also use `HandleJobStarted` here.

File: dashboard/modules/runtime_env/runtime_env_agent.py
Comment (SongGuyang) [reply dashboard/modules/runtime_env/runtime_env_agent.py:184]: I add `runtime_env` here because I don't want to deserialize twice. Just an optimization for performance.

File: src/ray/raylet/worker_pool.cc
Comment (SongGuyang) [reply src/ray/raylet/worker_pool.cc:139]: I think it applies. Here we kill all the worker processes when the worker pool is deconstructed. I didn't modify this logic. But I have redefine `starting_worker_processes` to `worker_processes` which means that we will store the alive processes all the way. So we can kill all the processes by this.

File: src/ray/raylet/worker_pool.cc
Comment (SongGuyang) [reply src/ray/raylet/worker_pool.cc:582]: Good! I will reuse yours.

File: src/ray/raylet/worker_pool.h
Comment (SongGuyang) [reply src/ray/raylet/worker_pool.h:841]: Yeah, I can add a doc to `Increase/DecreaseURIReference methods`. But it will be more simple which is entire different from this.

Comment (SongGuyang) [reply src/ray/raylet/worker_pool.h:841]: Add the doc here https://github.com/ray-project/ray/pull/22828/files#diff-a88735bc6aa09dd8248a3e73081916b5096a79c4bb7de91f9875d2c5a84ef075R629

File: dashboard/modules/runtime_env/runtime_env_agent.py
Comment (rkooo567) [inline dashboard/modules/runtime_env/runtime_env_agent.py:322]: The python code with try/except seems to make it difficult to make correct atomic ref counting. Why don't we refactor code in this way?

```python3
increase_ref_count(runtime_env)
status = create_runtime_env(runtime_env)
if status != success:
    decrease_ref_count(runtime_env)

reply(status)
    

```

File: dashboard/modules/runtime_env/runtime_env_agent.py
Comment (rkooo567) [inline dashboard/modules/runtime_env/runtime_env_agent.py:388]: Same comment as above?

File: dashboard/modules/runtime_env/runtime_env_agent.py
Comment (rkooo567) [reply dashboard/modules/runtime_env/runtime_env_agent.py:131]: We can also expose some gRPC endpoints to query ref count information in the future.

File: src/ray/raylet/worker_pool.cc
Comment (rkooo567) [inline src/ray/raylet/worker_pool.cc:189]: What's the motivation for the name change?

File: src/ray/raylet/worker_pool.h
Comment (rkooo567) [inline src/ray/raylet/worker_pool.h:634]: Can you mention this happens from the agent now?

Comment (rkooo567) [reply src/ray/raylet/worker_pool.h:634]: actual ref counting

File: src/ray/raylet/worker_pool.cc
Comment (rkooo567) [inline src/ray/raylet/worker_pool.cc:959]: Can `num_starting_workers ` be negative?

File: src/ray/raylet/worker_pool.cc
Comment (rkooo567) [inline src/ray/raylet/worker_pool.cc:1264]: Is it impossible the worker disconnected was a "starting workers"?

Comment (rkooo567) [reply src/ray/raylet/worker_pool.cc:1264]: It looks like this code could be problematic if that's the case

File: src/ray/raylet/worker_pool.cc
Comment (rkooo567) [inline src/ray/raylet/worker_pool.cc:1134]: Actually, why don't we call it `GetOrCreateRuntimeEnv`?

File: src/ray/raylet/worker_pool.cc
Comment (rkooo567) [inline src/ray/raylet/worker_pool.cc:1266]: For this API, my impression is the right name is 

```
DeleteRuntimeEnvIfPossible
```

or 
```
MarkRuntimeEnvUnused
```
Wdyt?

**Review by rkooo567 — COMMENTED** [review]
I had the initial pass! One of concerns I have is that gRPC requests are usually unordered afaik, and this could cause issues? (If Delete request comes before Create). Do you know if this is possible when the caller is always the same?

File: src/ray/raylet/worker_pool.cc
Comment (SongGuyang) [reply src/ray/raylet/worker_pool.cc:189]: Before current PR, we will delete the process when the worker finishes registration. 
In current PR, I change this logic. The process will be stored all the way before the worker exits.

**Comment by SongGuyang** [comment]
> I had the initial pass! One of concerns I have is that gRPC requests are usually unordered afaik, and this could cause issues? (If Delete request comes before Create). Do you know if this is possible when the caller is always the same?

TCP can guarantee the order, I'm not sure if grpc local buffer will break this(I guess it won't). But I think there is no consistency issue even though the Delete request comes before Create. If the Delete request comes first, agent will mark the runtime env unused and put it into the cache. When Create request comes, if the cache is not evicted, the runtime env will be marked used. Otherwise, agent will setup the runtime env again. If will bring extra workload, but no serious.

File: src/ray/raylet/worker_pool.cc
Comment (SongGuyang) [reply src/ray/raylet/worker_pool.cc:959]: I think it can't. But I also could change this code to `it->second.num_starting_workers > 0`.

File: src/ray/raylet/worker_pool.h
Comment (SongGuyang) [reply src/ray/raylet/worker_pool.h:634]: ok

File: src/ray/raylet/worker_pool.cc
Comment (SongGuyang) [reply src/ray/raylet/worker_pool.cc:1264]: Good point. I think this could happen by low probability. I have deleted this field and add a field `alive_started_workers` to record the started workers.

File: src/ray/raylet/worker_pool.cc
Comment (SongGuyang) [reply src/ray/raylet/worker_pool.cc:1266]: `CreateRuntimeEnvOrGet` and `DeleteRuntimeEnvIfPossible`? I like symmetrical names. 😄

**Comment by SongGuyang** [comment]
@rkooo567 @architkulkarni Please take a look again. I did a little refactor of runtime env agent.

File: src/ray/raylet/worker_pool.cc
Comment (rkooo567) [reply src/ray/raylet/worker_pool.cc]: Hmm imo these two has a slightly different semantics to have symmetric names (create is get Or create , and delete it Delete or do nothing). But it is up to you.

**Comment by kfstorm** [comment]
## :bangbang: ACTION REQUIRED :bangbang:

We've updated our formatting configuration for C++ code. (see #22725)

This PR includes C++ code change. To prevent issues with merging your code, here's what you'll need to do:

1. Merge the latest changes from upstream/master branch into your branch.

```bash
git pull upstream master
git merge upstream/master
```

2. Resolve merge conflicts (if necessary).

After running these steps, you'll have the updated C++ formatting configuration.

3. Format changed files.

```bash
scripts/format.sh
```

4. Commit your changes.

```bash
git add --all
git commit -m "Format C++ code"
```

**Comment by SongGuyang** [comment]
@architkulkarni I found that the bad runtime env result is not deleted forever in agent. I added a config namd `BAD_RUNTIME_ENV_CACHE_TTL_SECONDS ` to control this. Please check if this is reasonable https://github.com/ray-project/ray/pull/22828/commits/17872c1423dabf4462e7304db0863c86510c6121

**Review by architkulkarni — APPROVED** [review]
The reference table refactor looks great, thanks for doing it!  @SongGuyang In the future I think we can get rid of the runtime_env reference counting now that we have URI reference counting, right?  Or is the runtime_env reference counting still used for something?

**Comment by SongGuyang** [comment]
> @SongGuyang In the future I think we can get rid of the runtime_env reference counting now that we have URI reference counting, right? Or is the runtime_env reference counting still used for something?

I add runtime_env reference because I want to remove `_uris_to_envs` which seems a little complex. In future, maybe we can merge runtime_env reference with `_env_cache`? Or just keep it, I think it's also reasonable.

**Comment by SongGuyang** [comment]
@rkooo567 Any other comments?

**Comment by rkooo567** [comment]
I will make sure to review this by today!

**Comment by SongGuyang** [comment]
> I will make sure to review this by today!

OK. Looking forward. 😁

File: dashboard/modules/runtime_env/runtime_env_agent.py
Comment (rkooo567) [inline dashboard/modules/runtime_env/runtime_env_agent.py:85]: Do we need it? Isn't uri reference enough?

File: dashboard/modules/runtime_env/runtime_env_agent.py
Comment (rkooo567) [inline dashboard/modules/runtime_env/runtime_env_agent.py:84]: Consider `defaultdict(int)`?

File: dashboard/modules/runtime_env/runtime_env_agent.py
Comment (rkooo567) [inline dashboard/modules/runtime_env/runtime_env_agent.py:83]: Why do we have this self._logger btw? Isn't using default_logger sufficient?

File: dashboard/modules/runtime_env/runtime_env_agent.py
Comment (rkooo567) [inline dashboard/modules/runtime_env/runtime_env_agent.py:122]: same comment as defaultdict

File: dashboard/modules/runtime_env/runtime_env_agent.py
Comment (rkooo567) [inline dashboard/modules/runtime_env/runtime_env_agent.py:113]: Is it error? When does this usually happen?

File: dashboard/modules/runtime_env/runtime_env_agent.py
Comment (rkooo567) [inline dashboard/modules/runtime_env/runtime_env_agent.py:79]: Is it possible to make it not pass callable like this? Usually, this happens when the class hierarchy is badly structured. Is there any alternative option?

File: dashboard/modules/runtime_env/runtime_env_agent.py
Comment (rkooo567) [inline dashboard/modules/runtime_env/runtime_env_agent.py:263]: Any reason why it is not `GetOrCreateRuntimeEnv`?

File: dashboard/modules/runtime_env/runtime_env_agent.py
Comment (rkooo567) [inline dashboard/modules/runtime_env/runtime_env_agent.py:264]: Maybe debug?

File: dashboard/modules/runtime_env/runtime_env_agent.py
Comment (rkooo567) [inline dashboard/modules/runtime_env/runtime_env_agent.py:350]: Can you write a docstring?

**Review by rkooo567 — COMMENTED** [review]
Mostly looks good! Final NIT comments.

@architkulkarni would this change your refactoring plans (hide env_cache under the manager class)?

**Comment by architkulkarni** [comment]
> @architkulkarni would this change your refactoring plans (hide env_cache under the manager class)?
No, I think this makes it easier.  I can always refactor it more if needed

File: dashboard/modules/runtime_env/runtime_env_agent.py
Comment (SongGuyang) [reply dashboard/modules/runtime_env/runtime_env_agent.py:85]: As the comment https://github.com/ray-project/ray/pull/22828#issuecomment-1067532865 says, I think an additional `_runtime_env_reference` could make the logic more clear.

File: dashboard/modules/runtime_env/runtime_env_agent.py
Comment (SongGuyang) [reply dashboard/modules/runtime_env/runtime_env_agent.py:86]: This isn't a util function. `uris_parser` is a member function of `RuntimeEnvAgent` which can modify the states of class.

File: dashboard/modules/runtime_env/runtime_env_agent.py
Comment (SongGuyang) [reply dashboard/modules/runtime_env/runtime_env_agent.py:90]: It relies on job exiting. When job exits, the reference count becomes to zero.

File: dashboard/modules/runtime_env/runtime_env_agent.py
Comment (SongGuyang) [reply dashboard/modules/runtime_env/runtime_env_agent.py:83]: Just copy from `RuntimeEnvAgent`. I can remove this.

File: dashboard/modules/runtime_env/runtime_env_agent.py
Comment (SongGuyang) [reply dashboard/modules/runtime_env/runtime_env_agent.py:113]: It happens when the caller(raylet) sends redundant PRC to delete runtime env. I also can modify it to warn.

File: dashboard/modules/runtime_env/runtime_env_agent.py
Comment (SongGuyang) [reply dashboard/modules/runtime_env/runtime_env_agent.py:79]: Use callable here because `ReferenceTable` don't need to know the manager instance(_working_dir_manager, _pip_manager and so on) and the URI cache instance(_working_dir_uri_cache, _pip_uri_cache and so on). I agree that we can find a better way to refactor. But I think we could try it in future PR.

File: dashboard/modules/runtime_env/runtime_env_agent.py
Comment (SongGuyang) [reply dashboard/modules/runtime_env/runtime_env_agent.py:263]: Just make `CreateRuntimeEnv...` and `DeleteRuntimeEnv...` symmetrical. 😅

Comment (SongGuyang) [reply dashboard/modules/runtime_env/runtime_env_agent.py]: Rename done!

File: src/ray/raylet/agent_manager.cc
Comment (SongGuyang) [reply src/ray/raylet/agent_manager.cc]: Fixed after merge master.

**Comment by SongGuyang** [comment]
@rkooo567 Take a look again? We will merge today if no new comments.

File: dashboard/modules/runtime_env/runtime_env_agent.py
Comment (rkooo567) [reply dashboard/modules/runtime_env/runtime_env_agent.py]: Can you tell me when the redundant RPCs can be sent to agent? what's the exact scenario?

File: dashboard/modules/runtime_env/runtime_env_agent.py
Comment (rkooo567) [reply dashboard/modules/runtime_env/runtime_env_agent.py:80]: Hmm this makes me think all logic here should be just within `RuntimeEnvAgent` class. Having a separate class like this requires you to pass callbacks, which are weird. 

Can you just remove this class and make them a private method of `RuntimeEnvAgent`?

```python3
class RuntimeEnvAgent:
    def _increase_ref_count():
    def _decrease_ref_count():
```

File: dashboard/modules/runtime_env/runtime_env_agent.py
Comment (rkooo567) [inline dashboard/modules/runtime_env/runtime_env_agent.py:347]: Can you follow the proper Python-style docstring (https://docs.ray.io/en/master/ray-contribute/getting-involved.html#code-style)? Also, consider adding type annotation t

```python3
        async def _create_runtime_env_with_retry(
            runtime_env: RuntimeEnv, 
            serialized_runtime_env: str, 
            serialized_allocated_resource_instances: str
        ) -> Tuple[X, Y, Z]:
        """
         Blah blah
         
         Args:
         Returns:

        """
```

File: dashboard/modules/runtime_env/runtime_env_agent.py
Comment (rkooo567) [reply dashboard/modules/runtime_env/runtime_env_agent.py]: @architkulkarni why do we need uri_to_env at all? There's no meaning to follow env anymore right? We just need to track of uris here.

**Comment by rkooo567** [comment]
^ last 3 comments !

**Comment by SongGuyang** [comment]
@rkooo567 Do you remember that you comment here https://github.com/ray-project/ray/pull/22828#discussion_r822811144 to ask me using a single class? I think the current class is clear and the callback is reasonable. If you insist, I will add a `TODO` here and change it in future PR. 
btw, Only left one day for me to merge this PR. Can you approve first? I don't think the remaining comment issues are big.

**Review by rkooo567 — APPROVED** [review]

**Comment by SongGuyang** [comment]
I have pushed new commit here https://github.com/alipay/ray/commit/57f1bbc256c584bed8050d6d22a45c35e43abdf7 but current PR don't change.

**Comment by rkooo567** [comment]
https://github.com/ray-project/ray/pull/22828#issuecomment-1070978297 is this a bug? I heard from some people that Github has some outage now https://www.githubstatus.com/

**Comment by rkooo567** [comment]
If it takes time to resolve it, I can probably just merge it and you can create a follow up PR for the fix

**Comment by SongGuyang** [comment]
@rkooo567 I really appreciate for your careful review. This help me a lot to enhance my code. Respect for your pay out. But I don't think I'm producing tech debt. I also try to address or feed back all the comments carefully, right? If my PR don't meet the requirement of merging, no problem, I will improve.

About the class, I think the benefit is that it could make the responsibility clearly. `ReferenceTable` only maintains the reference count and handles others. If you think I lost the unit test, I'm ok to add.

**Comment by SongGuyang** [comment]
Add a reference table test https://github.com/ray-project/ray/pull/22828/commits/b80289edd3a97cb47ddb703c122720b14277f3c5.

**Review by rkooo567 — APPROVED** [review]
Looks like it has conflicts?

**Comment by SongGuyang** [comment]
> Looks like it has conflicts?

Yep, I merged master just now. Let's merge this if no tests broken.

**Comment by architkulkarni** [comment]
serve:test_cli flaky on master, Mac tests stalled

**Comment by SongGuyang** [comment]
@rkooo567 Is time to merge?

**Pull request merged**
